zypp-plugin/snapper_zypp_plugin.cc

```diff
@@ -0,0 +1,296 @@
+						      userdata);
+		y2deb("created pre snapshot " << pre_snapshot_num);
+	    }
+	    catch (const Exception& ex) {
+		// assumes a logging setup
+		SN_CAUGHT(ex);
+	    }
+	}
+
+	return ack();
+    }
+
+    Message commit_end(const Message& msg) override {
+	y2mil("COMMITEND");
+
+	if (pre_snapshot_num != 0) {
+	    set<string> solvables = get_solvables(msg, false);
+	    y2deb("solvables: " << solvables);
+
+	    bool found, important;
+	    match_solvables(solvables, found, important);
+	    y2mil("found: " << found << ", important: " << important);
+
+	    if (found || important) {
+		userdata["important"] = important ? "yes" : "no";
+
+		try {
+		    snapper::SMD modification_data;
+		    modification_data.description = snapshot_description;
+		    modification_data.cleanup = cleanup_algorithm;
+		    modification_data.userdata = userdata;
+		    command_set_snapshot(dbus_conn, "root", pre_snapshot_num, modification_data);
+		}
+		catch (const Exception& ex) {
+		    y2err("setting snapshot data failed:");
+		    // assumes a logging setup
+		    SN_CAUGHT(ex);
+		}
+		try {
+		    y2mil("creating post snapshot");
+		    unsigned int post_snapshot_num = command_create_post_snapshot(dbus_conn, "root", pre_snapshot_num, "", cleanup_algorithm,
+								     userdata);
+		    y2deb("created post snapshot " << post_snapshot_num);
+		}
+		catch (const Exception& ex) {
+		    y2err("creating snapshot failed:");
+		    // assumes a logging setup
+		    SN_CAUGHT(ex);
+		}
+	    }
+	    else {
+		try {
+		    y2mil("deleting pre snapshot");
+		    vector<unsigned int> nums{ pre_snapshot_num };
+		    bool verbose = false;
+		    command_delete_snapshots(dbus_conn, "root", nums, verbose);
+		    y2deb("deleted pre snapshot " << pre_snapshot_num);
+		}
+		catch (const Exception& ex) {
+		    y2err("deleting snapshot failed:");
+		    // assumes a logging setup
+		    SN_CAUGHT(ex);
+		}
+	    }
+	}
+	return ack();
+    }
+
+private:
+    static const string cleanup_algorithm;
+
+    DBus::Connection dbus_conn;
+    unsigned int pre_snapshot_num;
+    string snapshot_description;
+    map<string, string> userdata;
+
+    vector<SolvableMatcher> solvable_matchers;
+
+    map<string, string> get_userdata(const Message&);
+
+    // FIXME: what does the todo flag mean?
+    set<string> get_solvables(const Message&, bool todo);
+
+    void match_solvables(const set<string>& solvables, bool& found, bool& important);
+
+    unsigned int create_pre_snapshot(string config_name, string description, string cleanup, map<string, string> userdata);
+};
+
+const string SnapperZyppPlugin::cleanup_algorithm = "number";
+
+map<string, string> SnapperZyppPlugin::get_userdata(const Message&) {
+    map<string, string> result;
+    return result;
+}
+
+set<string> SnapperZyppPlugin::get_solvables(const Message& msg, bool todo) {
+    set<string> result;
+
+    rapidjson::Document doc;
```

Comment:
Umm, if we need a json library to import data, maybe we might modify the `JsonFormatter` implementation to use it.

Reply:
Oh, good point, I'll check that

Reply:
But maybe we should wait for Boost.JSON (https://github.com/vinniefalco/json). That should if accepted by available everywhere where boost is available (and that is required anyway).

Reply:
Currently I am linking to json-c where the *so* file has 64208 bytes. Some boost libraries have runtimes around a megabyte big, so we should check that factor too.

Reply:
Followup card: https://trello.com/c/L0GJlnET/3706-snapper-use-a-library-for-json-output